### ISSUE TITLE
NPM install of 'stampery' is broken, dependency issue with stampery/msgpackrpc

### ISSUE THREAD

**Issue by grempe** [stampery/node#2]
You require the `stampery/msgpackrpc` package from github.  Attempting to use `node` in a test pure npm/JS project results in dependency failures.

The `msgpackrpc` project does not contain JS code, only iced coffeescript.

```
~/src/stampery-test$ npm test

> stampery-test@1.0.0 test /Users/glenn/src/stampery-test
> node index.js

module.js:339
    throw err;
    ^


Error: Cannot find module 'msgpackrpc'
    at Function.Module._resolveFilename (module.js:337:15)
    at Function.Module._load (module.js:287:25)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/glenn/src/stampery-test/node_modules/stampery/index
.js:16:16)
    at Object.<anonymous> (/Users/glenn/src/stampery-test/node_modules/stampery/index
.js:390:4)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
npm ERR! Test failed.  See above for more details.
```

**Comment by MiquelGatiusF** [comment]
Hello found the solution for your issue:
package.json
.......................
  "dependencies": {
    "express": "^4.13.4",
    "stampery": "3.0.0",
    "msgpackrpc": "1.0.3"
    }
...........................

**Comment by aesedepece** [comment]
Hi, version 3.1.0 solves this issue and adds convenience methods for independent verification of proofs directly from the API.

Sorry for the inconveniences and thanks for your support.

**Comment by grempe** [comment]
Sorry, but I don't think this has resolved the issue.

In a new directory I have the following:

https://gist.github.com/grempe/3cae448e211ec7fe396c748e089f853e

Run this to install the latest `3.1.2` version of stampery/node:

```
~/src/stampery-test$ npm install stampery --save
npm WARN prefer global coffee-script@1.10.0 should be installed with -g

> msgpack@1.0.2 install /Users/glenn/src/stampery-test/node_modules/msgpack
> node-gyp rebuild

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
  CC(target) Release/obj.target/libmsgpack/deps/msgpack/objectc.o
  CC(target) Release/obj.target/libmsgpack/deps/msgpack/unpack.o
  CC(target) Release/obj.target/libmsgpack/deps/msgpack/vrefbuffer.o
  CC(target) Release/obj.target/libmsgpack/deps/msgpack/zone.o
  CC(target) Release/obj.target/libmsgpack/deps/msgpack/version.o
  LIBTOOL-STATIC Release/msgpack.a
  CXX(target) Release/obj.target/msgpackBinding/src/msgpack.o
  SOLINK_MODULE(target) Release/msgpackBinding.node

> sha3@1.2.0 install /Users/glenn/src/stampery-test/node_modules/sha3
> node-gyp rebuild

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
  CXX(target) Release/obj.target/sha3/src/addon.o
  CXX(target) Release/obj.target/sha3/src/displayIntermediateValues.o
  CXX(target) Release/obj.target/sha3/src/KeccakF-1600-reference.o
  CXX(target) Release/obj.target/sha3/src/KeccakNISTInterface.o
  CXX(target) Release/obj.target/sha3/src/KeccakSponge.o
  SOLINK_MODULE(target) Release/sha3.node
stampery-test@1.0.0 /Users/glenn/src/stampery-test
└─┬ stampery@3.1.2
  ├─┬ amqplib@0.4.2
  │ ├── bitsyntax@0.0.4
  │ ├── buffer-more-ints@0.0.2
  │ ├─┬ readable-stream@1.1.14
  │ │ ├── core-util-is@1.0.2
  │ │ ├── inherits@2.0.1
  │ │ ├── isarray@0.0.1
  │ │ └── string_decoder@0.10.31
  │ └── when@3.6.4
  ├─┬ grunt@1.0.1
  │ ├── coffee-script@1.10.0
  │ ├─┬ dateformat@1.0.12
  │ │ ├── get-stdin@4.0.1
  │ │ └─┬ meow@3.7.0
  │ │   ├─┬ camelcase-keys@2.1.0
  │ │   │ └── camelcase@2.1.1
  │ │   ├── decamelize@1.2.0
  │ │   ├─┬ loud-rejection@1.5.0
  │ │   │ ├─┬ currently-unhandled@0.4.1
  │ │   │ │ └── array-find-index@1.0.1
  │ │   │ └── signal-exit@3.0.0
  │ │   ├── map-obj@1.0.1
  │ │   ├── minimist@1.2.0
  │ │   ├─┬ normalize-package-data@2.3.5
  │ │   │ ├── hosted-git-info@2.1.5
  │ │   │ ├─┬ is-builtin-module@1.0.0
  │ │   │ │ └── builtin-modules@1.1.1
  │ │   │ ├── semver@5.1.0
  │ │   │ └─┬ validate-npm-package-license@3.0.1
  │ │   │   ├─┬ spdx-correct@1.0.2
  │ │   │   │ └── spdx-license-ids@1.2.1
  │ │   │   └─┬ spdx-expression-parse@1.0.2
  │ │   │     └── spdx-exceptions@1.0.4
  │ │   ├── object-assign@4.1.0
  │ │   ├─┬ read-pkg-up@1.0.1
  │ │   │ ├─┬ find-up@1.1.2
  │ │   │ │ └── path-exists@2.1.0
  │ │   │ └─┬ read-pkg@1.1.0
  │ │   │   ├─┬ load-json-file@1.1.0
  │ │   │   │ ├── graceful-fs@4.1.4
  │ │   │   │ ├─┬ parse-json@2.2.0
  │ │   │   │ │ └─┬ error-ex@1.3.0
  │ │   │   │ │   └── is-arrayish@0.2.1
  │ │   │   │ ├── pify@2.3.0
  │ │   │   │ └─┬ strip-bom@2.0.0
  │ │   │   │   └── is-utf8@0.2.1
  │ │   │   └── path-type@1.1.0
  │ │   ├─┬ redent@1.0.0
  │ │   │ ├─┬ indent-string@2.1.0
  │ │   │ │ └─┬ repeating@2.0.1
  │ │   │ │   └─┬ is-finite@1.0.1
  │ │   │ │     └── number-is-nan@1.0.0
  │ │   │ └── strip-indent@1.0.1
  │ │   └── trim-newlines@1.0.0
  │ ├── eventemitter2@0.4.14
  │ ├── exit@0.1.2
  │ ├─┬ findup-sync@0.3.0
  │ │ └── glob@5.0.15
  │ ├─┬ glob@7.0.5
  │ │ ├── fs.realpath@1.0.0
  │ │ ├─┬ inflight@1.0.5
  │ │ │ └── wrappy@1.0.2
  │ │ └── once@1.3.3
  │ ├─┬ grunt-cli@1.2.0
  │ │ └── resolve@1.1.7
  │ ├── grunt-known-options@1.1.0
  │ ├─┬ grunt-legacy-log@1.0.0
  │ │ ├── colors@1.1.2
  │ │ ├─┬ grunt-legacy-log-utils@1.0.0
  │ │ │ └── lodash@4.3.0
  │ │ ├── hooker@0.2.3
  │ │ ├── lodash@3.10.1
  │ │ └── underscore.string@3.2.3
  │ ├─┬ grunt-legacy-util@1.0.0
  │ │ ├── async@1.5.2
  │ │ ├── getobject@0.1.0
  │ │ ├── lodash@4.3.0
  │ │ └─┬ which@1.2.10
  │ │   └── isexe@1.1.2
  │ ├── iconv-lite@0.4.13
  │ ├─┬ js-yaml@3.5.5
  │ │ ├─┬ argparse@1.0.7
  │ │ │ └── sprintf-js@1.0.3
  │ │ └── esprima@2.7.2
  │ ├─┬ minimatch@3.0.2
  │ │ └─┬ brace-expansion@1.1.5
  │ │   ├── balanced-match@0.4.1
  │ │   └── concat-map@0.0.1
  │ ├─┬ nopt@3.0.6
  │ │ └── abbrev@1.0.9
  │ ├── path-is-absolute@1.0.0
  │ └── rimraf@2.2.8
  ├── iced-runtime@1.0.3
  ├─┬ msgpack@1.0.2
  │ └── nan@2.3.5
  ├── msgpackrpc@1.0.5  (git://github.com/stampery/msgpackrpc.git#5d1733864734ba19b5a71696de57a37195acee41)
  ├─┬ request@2.72.0
  │ ├── aws-sign2@0.6.0
  │ ├── aws4@1.4.1
  │ ├─┬ bl@1.1.2
  │ │ └─┬ readable-stream@2.0.6
  │ │   ├── isarray@1.0.0
  │ │   ├── process-nextick-args@1.0.7
  │ │   └── util-deprecate@1.0.2
  │ ├── caseless@0.11.0
  │ ├─┬ combined-stream@1.0.5
  │ │ └── delayed-stream@1.0.0
  │ ├── extend@3.0.0
  │ ├── forever-agent@0.6.1
  │ ├── form-data@1.0.0-rc4
  │ ├─┬ har-validator@2.0.6
  │ │ ├─┬ chalk@1.1.3
  │ │ │ ├── ansi-styles@2.2.1
  │ │ │ ├── escape-string-regexp@1.0.5
  │ │ │ ├─┬ has-ansi@2.0.0
  │ │ │ │ └── ansi-regex@2.0.0
  │ │ │ ├── strip-ansi@3.0.1
  │ │ │ └── supports-color@2.0.0
  │ │ ├─┬ commander@2.9.0
  │ │ │ └── graceful-readlink@1.0.1
  │ │ ├─┬ is-my-json-valid@2.13.1
  │ │ │ ├── generate-function@2.0.0
  │ │ │ ├─┬ generate-object-property@1.2.0
  │ │ │ │ └── is-property@1.0.2
  │ │ │ ├── jsonpointer@2.0.0
  │ │ │ └── xtend@4.0.1
  │ │ └─┬ pinkie-promise@2.0.1
  │ │   └── pinkie@2.0.4
  │ ├─┬ hawk@3.1.3
  │ │ ├── boom@2.10.1
  │ │ ├── cryptiles@2.0.5
  │ │ ├── hoek@2.16.3
  │ │ └── sntp@1.0.9
  │ ├─┬ http-signature@1.1.1
  │ │ ├── assert-plus@0.2.0
  │ │ ├─┬ jsprim@1.2.2
  │ │ │ ├── extsprintf@1.0.2
  │ │ │ ├── json-schema@0.2.2
  │ │ │ └── verror@1.3.6
  │ │ └─┬ sshpk@1.8.3
  │ │   ├── asn1@0.2.3
  │ │   ├── assert-plus@1.0.0
  │ │   ├─┬ dashdash@1.14.0
  │ │   │ └── assert-plus@1.0.0
  │ │   ├── ecc-jsbn@0.1.1
  │ │   ├─┬ getpass@0.1.6
  │ │   │ └── assert-plus@1.0.0
  │ │   ├── jodid25519@1.0.2
  │ │   ├── jsbn@0.1.0
  │ │   └── tweetnacl@0.13.3
  │ ├── is-typedarray@1.0.0
  │ ├── isstream@0.1.2
  │ ├── json-stringify-safe@5.0.1
  │ ├─┬ mime-types@2.1.11
  │ │ └── mime-db@1.23.0
  │ ├── node-uuid@1.4.7
  │ ├── oauth-sign@0.8.2
  │ ├── qs@6.1.0
  │ ├── stringstream@0.0.5
  │ ├── tough-cookie@2.2.2
  │ └── tunnel-agent@0.4.3
  ├── retry@0.8.0
  ├── rocksolidsocket@1.0.0
  └── sha3@1.2.0

npm WARN stampery-test@1.0.0 No description
npm WARN stampery-test@1.0.0 No repository field.
```

Note: `msgpackrpc` is being installed from the head of the master branch of your git repository:

```
...
git://github.com/stampery/msgpackrpc.git#5d1733864734ba19b5a71696de57a37195acee41
...
```

Trying to run the example script results in:

```
~/src/stampery-test$ node index.js
module.js:339
    throw err;
    ^

Error: Cannot find module 'msgpackrpc'
    at Function.Module._resolveFilename (module.js:337:15)
    at Function.Module._load (module.js:287:25)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/glenn/src/stampery-test/node_modules/stampery/index.js:16:16)
    at Object.<anonymous> (/Users/glenn/src/stampery-test/node_modules/stampery/index.js:518:4)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
```

Note: It cannot load the msgpackrpc module.

I believe this is due to the fact that your custom msgpackrpc module specifies a `main` file of `index.js`.

https://github.com/stampery/msgpackrpc/blob/master/package.json#L5

This is a file that **does not exist**.  In fact there is no JavaScript in that npm module at all.  There is only an `index.iced` file.

NPM Packages are for JavaScript, not for any flavor of CoffeeScript.  Neither Node, nor the Browser know how to used an `*.iced` file. Especially one that is not even referenced in the `package.json` file.

http://stackoverflow.com/questions/13645824/can-i-write-npm-package-in-coffeescript

I believe this is why this is failing.  In doing your development I can only surmise that you have actually compiled your `index.iced` into an `index.js` file in your `msgpackrpc` dir.

Also, it is considered poor form to reference the head of a git repository as a dependency.  You should reference a specific commit or tag so all of your users are not pulling in breaking changes.  The even better solution would be to not reference a git repo at all and instead to publish a working version of `msgpackrpc` to npm.

Additionally, you have published the `stampery` NPM package `v3.1.2` to npm:

https://www.npmjs.com/package/stampery

However, you have apparently published this from a `newapi` branch of your Github repository which is considered poor form.  Additionally, the `newapi` branch is in conflict with the `master` as you have conflicting commits on each.  If this is the new code, you would be encouraged to merge `newapi` onto master, resolve the conflicts, and publish a new version of the package to npm from that newly updated `master`.

I'll try again when you have resolved these issues.

**Comment by grempe** [comment]
Please re-open this issue.  It is not resolved. See comments above.
